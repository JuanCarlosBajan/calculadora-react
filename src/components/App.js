import React, { useState } from 'react'
import infixToPostfix from 'infix-to-postfix'
import postfixCalculator from 'postfix-calculator'
import '../styles/Button.css'
import '../styles/Answer.css'
import '../styles/App.css'

export default function App() {
  const [answer, setAnswer] = useState('')
  const [expression, setExpression] = useState('')
  const [actual, setActual] = useState('')
  const [toggleSign, setToggleSign] = useState(false)

  function isIn(value) {
    if (
      value.includes(' + ') ||
      value.includes(' - ') ||
      value.includes(' * ') ||
      value.includes(' / ')
    )
      return true
    return false
  }

  function verifyAnswer(value) {
    if (answer && !isIn(expression.toString())) {
      setAnswer('')
      setExpression('')
    }
    // eslint-disable-next-line yoda
    if (actual.length < 9) {
      setActual(actual + value)
    }
  }

  function addDot() {
    if (actual) {
      verifyAnswer('.')
    } else {
      verifyAnswer('0.')
    }
  }

  function del() {
    if (actual) {
      setActual(actual.substring(0, actual.length - 1))
    }

    if (actual.length === 2 && toggleSign) {
      setActual('')
      setToggleSign(!toggleSign)
    }
  }

  function clear() {
    setActual('')
    setExpression('')
    setAnswer('')
  }

  function sign() {
    if (!toggleSign && actual) {
      setActual('-' + actual)
      setToggleSign(!toggleSign)
    } else if (toggleSign && actual) {
      setActual(actual.substring(1))
      setToggleSign(!toggleSign)
    }
  }

  function equals() {
    if (expression && actual) {
      let response = postfixCalculator(infixToPostfix(expression + actual))
      clear()
      if (response > 999999999 || response < -999999999) {
        setAnswer('ERROR')
        setExpression('')
      } else {
        if (response.toString().length > 9) {
          response = response.toString().substring(0, 9)
        }
        setAnswer(response)
        setExpression(response)
      }
      setActual('')
    }
  }

  function addExpression(value) {
    if (actual) {
      setExpression(expression + actual + ' ' + value + ' ')
      setActual('')
    } else if (!actual && expression) {
      setExpression(expression + ' ' + value + ' ')
    }
  }

  return (
    <div className="main">
      <h1>ANSWER: {answer}</h1>
      <h1 id="value">VALUE: {actual}</h1>
      <div>
        <button onClick={() => clear()}>Clear</button>
        <button onClick={() => sign()}>+/-</button>
        <button onClick={() => addExpression('')}>_</button>
        <button onClick={() => addExpression('/')}>/</button>
      </div>
      <div>
        <button onClick={() => verifyAnswer('1')}>1</button>
        <button onClick={() => verifyAnswer('2')}>2</button>
        <button onClick={() => verifyAnswer('3')}>3</button>
        <button onClick={() => addExpression('*')}>x</button>
      </div>
      <div>
        <button onClick={() => verifyAnswer('4')}>4</button>
        <button onClick={() => verifyAnswer('5')}>5</button>
        <button onClick={() => verifyAnswer('6')}>6</button>
        <button onClick={() => addExpression('-')}>-</button>
      </div>
      <div>
        <button onClick={() => verifyAnswer('7')}>7</button>
        <button onClick={() => verifyAnswer('8')}>8</button>
        <button onClick={() => verifyAnswer('9')}>9</button>
        <button onClick={() => addExpression('+')}>+</button>
      </div>
      <div>
        <button onClick={() => verifyAnswer('0')}>0</button>
        <button onClick={() => addDot()}>.</button>
        <button onClick={() => del()}>{'<'}</button>
        <button onClick={() => equals()}>=</button>
      </div>
    </div>
  )
}
