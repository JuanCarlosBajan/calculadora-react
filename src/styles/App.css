body {
  background-color: #e4e4e4;
  height: 100vh;
  display: flex;
  align-items: center;
  justify-content: center;
}
