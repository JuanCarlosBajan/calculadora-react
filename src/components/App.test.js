/* eslint-disable no-undef */
import { fireEvent } from '@testing-library/react'
import React from 'react'
import ReactDOM from 'react-dom/client'
import { act } from 'react-dom/test-utils'
import renderer from 'react-test-renderer'
//import { render, fireEvent, getByTestId } from 'react-testing-library'
import App from './App'

let container

beforeEach(() => {
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  document.body.removeChild(container)
  container = null
})

test('Renders without crashing', () => {
  act(() => {
    ReactDOM.createRoot(container).render(<App />)
  })
})

test('Funcionamiento al seleccionar 1, 2, 3 y *', () => {
  act(() => {
    ReactDOM.createRoot(container).render(<App />)
  })
  const labels = container.querySelectorAll('h1')
  const btns = container.querySelectorAll('button')

  function click(element) {
    fireEvent(
      element,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
  }
  click(btns[4])
  expect(labels[1].textContent).toBe('VALUE: 1')
  click(btns[5])
  expect(labels[1].textContent).toBe('VALUE: 12')
  click(btns[6])
  expect(labels[1].textContent).toBe('VALUE: 123')
  click(btns[7])
  expect(labels[1].textContent).toBe('VALUE: ')
})

test('Funcionamiento al seleccionar 4, 5, 6 y -', () => {
  act(() => {
    ReactDOM.createRoot(container).render(<App />)
  })
  const labels = container.querySelectorAll('h1')
  const btns = container.querySelectorAll('button')

  function click(element) {
    fireEvent(
      element,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
  }
  click(btns[8])
  expect(labels[1].textContent).toBe('VALUE: 4')
  click(btns[9])
  expect(labels[1].textContent).toBe('VALUE: 45')
  click(btns[10])
  expect(labels[1].textContent).toBe('VALUE: 456')
  click(btns[11])
  expect(labels[1].textContent).toBe('VALUE: ')
})

test('Funcionamiento al seleccionar 7, 8, 9 y +', () => {
  act(() => {
    ReactDOM.createRoot(container).render(<App />)
  })
  const labels = container.querySelectorAll('h1')
  const btns = container.querySelectorAll('button')

  function click(element) {
    fireEvent(
      element,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
  }
  click(btns[12])
  expect(labels[1].textContent).toBe('VALUE: 7')
  click(btns[13])
  expect(labels[1].textContent).toBe('VALUE: 78')
  click(btns[14])
  expect(labels[1].textContent).toBe('VALUE: 789')
  click(btns[15])
  expect(labels[1].textContent).toBe('VALUE: ')
})

test('Renders correctly with snapshot', () => {
  const tree = renderer.create(<App />).toJSON()
  expect(tree).toMatchSnapshot()
})

test('Funcionamiento de toggle sign +/-', () => {
  act(() => {
    ReactDOM.createRoot(container).render(<App />)
  })
  const labels = container.querySelectorAll('h1')
  const btns = container.querySelectorAll('button')

  function click(element) {
    fireEvent(
      element,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
  }
  click(btns[12])
  expect(labels[1].textContent).toBe('VALUE: 7')
  click(btns[1])
  expect(labels[1].textContent).toBe('VALUE: -7')
  click(btns[1])
  expect(labels[1].textContent).toBe('VALUE: 7')
})

test('Funcionamiento = (respuesta correcta)', () => {
  act(() => {
    ReactDOM.createRoot(container).render(<App />)
  })
  const labels = container.querySelectorAll('h1')
  const btns = container.querySelectorAll('button')

  function click(element) {
    fireEvent(
      element,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
  }
  click(btns[12])
  expect(labels[1].textContent).toBe('VALUE: 7')
  click(btns[13])
  expect(labels[1].textContent).toBe('VALUE: 78')
  click(btns[14])
  expect(labels[1].textContent).toBe('VALUE: 789')
  click(btns[11])
  expect(labels[1].textContent).toBe('VALUE: ')
  click(btns[12])
  expect(labels[1].textContent).toBe('VALUE: 7')
  click(btns[13])
  expect(labels[1].textContent).toBe('VALUE: 78')
  click(btns[19])
  expect(labels[0].textContent).toBe('ANSWER: 711')
})

test('Funcionamiento Clear', () => {
  act(() => {
    ReactDOM.createRoot(container).render(<App />)
  })
  const labels = container.querySelectorAll('h1')
  const btns = container.querySelectorAll('button')

  function click(element) {
    fireEvent(
      element,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
  }
  click(btns[12])
  expect(labels[1].textContent).toBe('VALUE: 7')
  click(btns[13])
  expect(labels[1].textContent).toBe('VALUE: 78')
  click(btns[14])
  expect(labels[1].textContent).toBe('VALUE: 789')
  click(btns[11])
  expect(labels[1].textContent).toBe('VALUE: ')
  click(btns[12])
  expect(labels[1].textContent).toBe('VALUE: 7')
  click(btns[13])
  expect(labels[1].textContent).toBe('VALUE: 78')
  click(btns[19])
  expect(labels[0].textContent).toBe('ANSWER: 711')
  click(btns[12])
  expect(labels[1].textContent).toBe('VALUE: 7')
  click(btns[0])
  expect(labels[0].textContent).toBe('ANSWER: ')
  expect(labels[1].textContent).toBe('VALUE: ')
})

test('Verify if calculation generates ERROR', () => {
  act(() => {
    ReactDOM.createRoot(container).render(<App />)
  })
  const labels = container.querySelectorAll('h1')
  const btns = container.querySelectorAll('button')

  function click(element) {
    fireEvent(
      element,
      new MouseEvent('click', {
        bubbles: true,
        cancelable: true,
      })
    )
  }
  click(btns[14])
  click(btns[14])
  click(btns[14])
  click(btns[14])
  click(btns[14])
  click(btns[14])
  click(btns[14])
  click(btns[7])
  click(btns[4])
  click(btns[16])
  click(btns[16])
  click(btns[16])
  click(btns[16])
  click(btns[19])
  expect(labels[0].textContent).toBe('ANSWER: ERROR')
})
