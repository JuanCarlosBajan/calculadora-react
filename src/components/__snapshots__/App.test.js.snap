// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly with snapshot 1`] = `
<div
  className="main"
>
  <h1>
    ANSWER: 
  </h1>
  <h1
    id="value"
  >
    VALUE: 
  </h1>
  <div>
    <button
      onClick={[Function]}
    >
      Clear
    </button>
    <button
      onClick={[Function]}
    >
      +/-
    </button>
    <button
      onClick={[Function]}
    >
      _
    </button>
    <button
      onClick={[Function]}
    >
      /
    </button>
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      1
    </button>
    <button
      onClick={[Function]}
    >
      2
    </button>
    <button
      onClick={[Function]}
    >
      3
    </button>
    <button
      onClick={[Function]}
    >
      x
    </button>
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      4
    </button>
    <button
      onClick={[Function]}
    >
      5
    </button>
    <button
      onClick={[Function]}
    >
      6
    </button>
    <button
      onClick={[Function]}
    >
      -
    </button>
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      7
    </button>
    <button
      onClick={[Function]}
    >
      8
    </button>
    <button
      onClick={[Function]}
    >
      9
    </button>
    <button
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <div>
    <button
      onClick={[Function]}
    >
      0
    </button>
    <button
      onClick={[Function]}
    >
      .
    </button>
    <button
      onClick={[Function]}
    >
      &lt;
    </button>
    <button
      onClick={[Function]}
    >
      =
    </button>
  </div>
</div>
`;
